# Module 6 - Challenge

In [1]:
# Import the dependencies.
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import requests
from config import weather_api_key
from datetime import datetime
import time
from scipy.stats import linregress

# Part 1

In [2]:
# Create a set of random latitude and longitude combinations.
lats = np.random.uniform(low=-90.000, high=90.000, size=1500)
lngs = np.random.uniform(low=-180.000, high=180.000, size=1500)
lat_lngs = zip(lats, lngs)
lat_lngs 

In [3]:
# Add the latitudes and longitudes to a list.
coordinates = list(lat_lngs)

In [4]:
# Use the citipy module to determine city based on latitude and longitude.
from citipy import citipy

In [5]:
# Create a list for holding the cities.
cities = []
# Identify the nearest city for each latitude and longitude combination.
for coordinate in coordinates:
    city = citipy.nearest_city(coordinate[0], coordinate[1]).city_name
    
    # If the city is unique, then we will add it to the cities list.
    if city not in cities:
        cities.append(city)
# Print the city count to confirm sufficient count.
len(cities)

600

In [6]:
# Import the API key.
from config import weather_api_key

In [12]:
cities

['mehamn',
 'kawhia',
 'puerto ayora',
 'hidrolandia',
 'castro',
 'busselton',
 'portland',
 'kapaa',
 'taolanaro',
 'jian',
 'alofi',
 'yellowknife',
 'matam',
 'port pirie',
 'tuktoyaktuk',
 'hobart',
 'punta arenas',
 'broome',
 'san policarpo',
 'guerrero negro',
 'yenisehir',
 'ushuaia',
 'syriam',
 'bredasdorp',
 'narsaq',
 'pedernales',
 'hilo',
 'san cristobal',
 'thompson',
 'mataura',
 'hermanus',
 'brzeg',
 'bambous virieux',
 'rikitea',
 'new norfolk',
 'shingu',
 'carnarvon',
 'mar del plata',
 'qaqortoq',
 'tabou',
 'nanortalik',
 'lebu',
 'cape town',
 'coihaique',
 'jamestown',
 'saint george',
 'san patricio',
 'blagoyevo',
 'amderma',
 'kamenskoye',
 'port alfred',
 'ponta do sol',
 'lompoc',
 'mahebourg',
 'albany',
 'tasiilaq',
 'luderitz',
 'bagotville',
 'xinzhou',
 'cajaruro',
 'airai',
 'barrow',
 'karratha',
 'dunedin',
 'bluff',
 'moissala',
 'ust-ilimsk',
 'bethel',
 'kulhudhuffushi',
 'lorengau',
 'ribeira brava',
 'sangar',
 'grand river south east',
 'sev

In [7]:
# Starting URL for Weather Map API Call.
url = "http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=" + "4ce7e5c77188098a3f13b86ba740d9ef"
print(url)

http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=4ce7e5c77188098a3f13b86ba740d9ef


In [32]:
# Create an empty list to hold the weather data.
city_data = []
# Print the beginning of the logging.
print("Beginning Data Retrieval     ")
print("-----------------------------")

# Create counters.
record_count = 1
set_count = 1
# Loop through all the cities in our list.
# Loop through all the cities in the list.
for i, city in enumerate(cities):

    # Group cities in sets of 50 for logging purposes.
    if (i % 50 == 0 and i >= 50):
        set_count += 1
        record_count = 1
    # Create endpoint URL with each city.
    city_url = url + "&q=" + city

    # Log the URL, record, and set numbers and the city.
    print(f"Processing Record {record_count} of Set {set_count} | {city}")
    # Add 1 to the record count.
    record_count += 1
    # Run an API request for each of the cities.
    try:
        # Parse the JSON and retrieve data.
        city_weather = requests.get(city_url).json()
        # Parse out the needed data.
        city_lat = city_weather["coord"]["lat"]
        city_country = city_weather["sys"]["country"]
        city_lng = city_weather["coord"]["lon"]
        city_max_temp = city_weather["main"]["temp_max"]
        city_min_temp = city_weather["main"]["temp_min"]
        city_humidity = city_weather["main"]["humidity"]
        city_clouds = city_weather["clouds"]["all"]
        city_wind = city_weather["wind"]["speed"]
        city_description = city_weather["weather"][0]["description"]
        try:
            city_rain = city_weather["rain"]["3h"]
        except:
            city_rain = 0
        try:
            city_snow = city_weather["snow"]["3h"]
        except:
            city_snow = 0
        # Convert the date to ISO standard.
        city_date = datetime.utcfromtimestamp(city_weather["dt"]).strftime('%Y-%m-%d %H:%M:%S')
        # Append the city information into city_data list.
        city_data.append({"City": city.title(),
                          "Country": city_country,
                          "Lat": city_lat,
                          "Lng": city_lng,
                          "Max Temp": city_max_temp,
                          "Min Temp": city_min_temp,
                          "Humidity": city_humidity,
                          "Cloudiness": city_clouds,
                          "Wind Speed": city_wind,
                          "Rain inches (past 3 hours)": city_rain,
                          "Snow inches (past 3 hours)": city_snow})
    # If an error is experienced, skip the city.
    except:
        print("City not found. Skipping...")
        pass

    # Indicate that Data Loading is complete.
    print("-----------------------------")
    print("Data Retrieval Complete      ")
    print("-----------------------------")

Beginning Data Retrieval     
-----------------------------
Processing Record 1 of Set 1 | mehamn
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 2 of Set 1 | kawhia
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 3 of Set 1 | puerto ayora
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 4 of Set 1 | hidrolandia
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 5 of Set 1 | castro
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 6 of Set 1 | busselton
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 7 of Set 1 | portland
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 8 of Set 1

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 15 of Set 2 | bluff
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 16 of Set 2 | moissala
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 17 of Set 2 | ust-ilimsk
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 18 of Set 2 | bethel
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 19 of Set 2 | kulhudhuffushi
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 20 of Set 2 | lorengau
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 21 of Set 2 | ribeira brava
-----------------------------
Data Retrieval Complete      
---------------

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 27 of Set 3 | hofn
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 28 of Set 3 | tondi
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 29 of Set 3 | port elizabeth
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 30 of Set 3 | salalah
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 31 of Set 3 | hai duong
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 32 of Set 3 | komsomolskiy
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 33 of Set 3 | natchitoches
-----------------------------
Data Retrieval Complete      
----------------

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 39 of Set 4 | tambovka
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 40 of Set 4 | buraydah
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 41 of Set 4 | ancud
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 42 of Set 4 | hun
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 43 of Set 4 | tiksi
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 44 of Set 4 | mayumba
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 45 of Set 4 | sabang
-----------------------------
Data Retrieval Complete      
-----------------------------
Process

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 1 of Set 6 | lugazi
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 2 of Set 6 | rovenki
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 3 of Set 6 | unai
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 4 of Set 6 | santa cruz
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 5 of Set 6 | guaruja
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 6 of Set 6 | mackay
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 7 of Set 6 | kormilovka
-----------------------------
Data Retrieval Complete      
-----------------------------
Proces

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 14 of Set 7 | palabuhanratu
City not found. Skipping...
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 15 of Set 7 | mayo
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 16 of Set 7 | provideniya
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 17 of Set 7 | linxia
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 18 of Set 7 | flinders
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 19 of Set 7 | eyl
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 20 of Set 7 | truckee
-----------------------------
Data Retrieval Complete      

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 26 of Set 8 | te anau
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 27 of Set 8 | kelo
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 28 of Set 8 | halalo
City not found. Skipping...
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 29 of Set 8 | alamosa
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 30 of Set 8 | nurota
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 31 of Set 8 | ugoofaaru
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 32 of Set 8 | cap malheureux
-----------------------------
Data Retrieval Complete     

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 38 of Set 9 | katsuura
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 39 of Set 9 | bijar
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 40 of Set 9 | bud
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 41 of Set 9 | tumannyy
City not found. Skipping...
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 42 of Set 9 | kedrovoye
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 43 of Set 9 | viligili
City not found. Skipping...
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 44 of Set 9 | udachnyy
-----------------------------
Data

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 50 of Set 10 | tarakan
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 1 of Set 11 | perth
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 2 of Set 11 | aloleng
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 3 of Set 11 | bima
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 4 of Set 11 | pevek
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 5 of Set 11 | saint anthony
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 6 of Set 11 | pindushi
-----------------------------
Data Retrieval Complete      
-----------------------------

-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 12 of Set 12 | amli
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 13 of Set 12 | lermontovka
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 14 of Set 12 | zhangye
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 15 of Set 12 | deh rawud
City not found. Skipping...
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 16 of Set 12 | santa rosalia
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 17 of Set 12 | valladolid
-----------------------------
Data Retrieval Complete      
-----------------------------
Processing Record 18 of Set 12 | sao joao da barra
-----------------------------
Dat

In [26]:
# Create an example of an endpoint URL for a city.
city_url = url + "&q=" + "fargo"
city_weather = requests.get(city_url)
city_weather.json()

{'coord': {'lon': -96.79, 'lat': 46.88},
 'weather': [{'id': 804,
   'main': 'Clouds',
   'description': 'overcast clouds',
   'icon': '04d'}],
 'base': 'stations',
 'main': {'temp': 38.66,
  'feels_like': 26.8,
  'temp_min': 37.99,
  'temp_max': 39.2,
  'pressure': 1019,
  'humidity': 80},
 'visibility': 16093,
 'wind': {'speed': 14.99, 'deg': 20},
 'clouds': {'all': 90},
 'dt': 1589063489,
 'sys': {'type': 1,
  'id': 4060,
  'country': 'US',
  'sunrise': 1589021952,
  'sunset': 1589075280},
 'timezone': -18000,
 'id': 5059163,
 'name': 'Fargo',
 'cod': 200}

In [33]:
# Convert the array of dictionaries to a Pandas DataFrame.
city_data_df = pd.DataFrame(city_data)
city_data_df.head(10)

City Country    Lat     Lng  Max Temp  Min Temp  Humidity  \
0        Mehamn      NO  71.04   27.85     29.88     29.88        90   
1        Kawhia      NZ -38.07  174.82     60.01     60.01        81   
2  Puerto Ayora      EC  -0.74  -90.35     84.20     79.00        70   
3   Hidrolandia      BR  -4.41  -40.44     73.18     73.18        80   
4        Castro      BR -24.79  -50.01     52.66     52.66        87   
5     Busselton      AU -33.65  115.33     39.00     39.00        75   
6      Portland      US  45.52 -122.68     87.01     81.00        23   
7         Kapaa      US  22.08 -159.32     80.60     77.00        61   
8          Jian      CN  27.12  114.98     72.07     72.07        93   
9         Alofi      NU -19.06 -169.92     84.20     84.20        79   

   Cloudiness  Wind Speed  Rain inches (past 3 hours)  \
0          93        7.00                           0   
1          37        1.01                           0   
2          20       20.80                           0   
3          20        2.93                           0   
4           0        3.53                           0   
5           0        2.95                           0   
6          75       18.34                           0   
7          40       17.22                           0   
8          95        6.82                           0   
9          67       10.29                           0   

   Snow inches (past 3 hours)  
0                           0  
1                           0  
2                           0  
3                           0  
4                           0  
5                           0  
6                           0  
7                           0  
8                           0  
9                           0

In [28]:
# Create the output file (CSV).
output_data_file = "weather_data/WeatherPy_challenge.csv"
# Export the City_Data into a CSV.
city_data_df.to_csv(output_data_file, index_label="City_ID")

In [29]:
len(city_data_df)

551

# How many cities have rain or snow?

# Answer = 551